pkg/util/atomic_copier.go

```diff
@@ -0,0 +1,132 @@
+package util
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+	"strings"
+
+	"k8s.io/klog/v2"
+)
+
+var (
+	cpArgs = []string{
+		"--verbose",
+		"--recursive",
+		"--preserve",
+		"--reflink=auto",
+	}
+)
+
+// AtomicCopier performs a two operation: copies source path to an intermediate location,
+// then renames from intermediate to final destination.
+// On Unix systems, the rename operation is atomic. This ensures that the file
+// is either fully renamed or not at all, which helps prevent issues like partial
+// file copies in cases of power failure or unexpected program termination.
+type AtomicCopier struct {
+	Source      string
+	Destination string
+
+	intermediatePath string
+}
+
+func (c *AtomicCopier) CopyToIntermediate() error {
+	if c == nil {
+		return nil
+	}
+	c.intermediatePath = fmt.Sprintf("%s.tmp", c.Destination)
+	if exists, err := PathExists(c.intermediatePath); err != nil {
+		return err
+	} else if exists {
+		if err := os.RemoveAll(c.intermediatePath); err != nil {
+			klog.Errorf("Failed to remove intermediate path %q which already existed: %v", c.intermediatePath, err)
+			return fmt.Errorf("failed to remove %q: %w", c.intermediatePath, err)
+		}
+	}
+	cmd := exec.Command("cp", append(cpArgs, c.Source, c.intermediatePath)...) //nolint:gosec
+
+	var outb, errb bytes.Buffer
+	cmd.Stdout = &outb
+	cmd.Stderr = &errb
+	err := cmd.Run()
+
+	if err != nil {
+		klog.ErrorS(nil, "Failed to make an intermediate copy", "cmd", cmd,
+			"stdout", strings.ReplaceAll(outb.String(), "\n", `, `),
+			"stderr", errb.String())
+		copyErr := fmt.Errorf("failed to copy %q to %q: %w", c.Source, c.intermediatePath, err)
+		if err := os.RemoveAll(c.intermediatePath); err != nil {
+			return errors.Join(copyErr, fmt.Errorf("failed to remove intermediate path %q: %w", c.intermediatePath, err))
+		}
+		return copyErr
+	}
+	klog.InfoS("Made an intermediate copy", "cmd", cmd)
+
+	return nil
+}
+
+func (c *AtomicCopier) RollbackIntermediate() error {
+	if c == nil {
+		return nil
+	}
+	if c.intermediatePath == "" {
+		return nil
+	}
+	if err := os.RemoveAll(c.intermediatePath); err != nil {
+		return fmt.Errorf("failed to remove intermediate path %q: %w", c.intermediatePath, err)
+	}
+	return nil
+}
+
+func (c *AtomicCopier) RenameToFinal() error {
+	if c == nil {
+		return nil
+	}
+	var src, dest string
+	if c.intermediatePath == "" {
+		// Empty value means there was no intermediate copy.
+		src = c.Source
+		dest = c.Destination
+	} else {
+		// Path was copied to a temporary location with .tmp suffix.
+		// Now it needs to be renamed into final destination.
+		// This two-step operation should provide a high guarantee that
+		// copying is complete and not partial thanks to rename being OS/filesystem atomic.
+		src = c.intermediatePath
+		dest = c.Destination
+	}
+
+	// Delete the destination if it's a non-empty directory.
+	// This is a limitation of the POSIX's rename.
+	if err := removeDirIfExists(dest); err != nil {
```

Comment:
I think this should be a fatal error. We should not delete leftover paths as a byproduct of copy operation.

Reply:
This is not leftover path - this is final path. And without this `mv /var/lib/microshift.tmp /var/lib/microshift` doesn't work because of `Rename()` limitations in POSIX and UNIX.

https://pubs.opengroup.org/onlinepubs/9699919799/functions/rename.html
> If new names an existing directory, it shall be required to be an empty directory.

```
$ mkdir 1 2; touch 1/a 2/b
$ mv 1 2
$ ls
2/
$ ls 2
1/  b
```
From `mv` man (same thing in the `cp` as they use the same library underneath):
```
       -T, --no-target-directory
              treat DEST as a normal file
```

```
$ mkdir 1 2; touch 1/a 2/b
$ mv -T 1 2
mv: cannot move '1' to '2': File exists
```